src/backend/opencl/kernel/random_philox.cl

```diff
@@ -0,0 +1,324 @@
+typedef uint  uint32_t;
+
+#define PI_VAL 3.1415926535897932384626433832795028841971693993751058209749445923078164
+#define m4x32_0                 0xD2511F53u
+#define m4x32_1                 0xCD9E8D57u
+#define w32_0                   0x9E3779B9u
+#define w32_1                   0xBB67AE85u
+#define PHILOX_DEFAULT_ROUNDS   10u
+#define R                       PHILOX_DEFAULT_ROUNDS
+#define UINTMAXFLOAT            4294967296.0f
+#define UINTLMAXDOUBLE          4294967296.0*4294967296.0
+
+struct philox_32_4_key_t
+{
+    uint v[2];
+};
+
+struct philox_32_4_ctr_t
+{
+    uint v[4];
+};
+
+void mulhilo(uint a, uint const * const b,
+    uint * const hi, uint * const lo)
+{
+    *hi = mul_hi(a, *b);
+    *lo = a*(*b);
+}
+
+void philox_round(struct philox_32_4_ctr_t * const ctr,
+    struct philox_32_4_key_t const * const key)
+{
+    uint hi0, lo0, hi1, lo1;
+    mulhilo(m4x32_0, &(ctr->v[0]), &hi0, &lo0);
+    mulhilo(m4x32_1, &(ctr->v[2]), &hi1, &lo1);
+    ctr->v[0] = hi1^(ctr->v[1])^(key->v[0]);
+    ctr->v[1] = lo1;
+    ctr->v[2] = hi0^(ctr->v[3])^(key->v[1]);
+    ctr->v[3] = lo0;
+}
+
+void philox_bump(struct philox_32_4_key_t * const key)
+{
+    key->v[0] += w32_0;
+    key->v[1] += w32_1;
+}
+
+struct philox_32_4_ctr_t philox(struct philox_32_4_ctr_t ctr,
+    struct philox_32_4_key_t key)
+{
+    struct philox_32_4_key_t const * const keyptr_read_only = (struct philox_32_4_key_t const * const) (&key);
+    struct philox_32_4_key_t * const keyptr = (struct philox_32_4_key_t * const) (&key);
+    struct philox_32_4_ctr_t * const ctrptr = (struct philox_32_4_ctr_t * const) (&ctr);
+#if R > 0
+    philox_round(ctrptr, keyptr_read_only);
+#endif
+#if R > 1
+    philox_bump(keyptr); philox_round(ctrptr, keyptr_read_only);
+#endif
+#if R > 2
+    philox_bump(keyptr); philox_round(ctrptr, keyptr_read_only);
+#endif
+#if R > 3
+    philox_bump(keyptr); philox_round(ctrptr, keyptr_read_only);
+#endif
+#if R > 4
+    philox_bump(keyptr); philox_round(ctrptr, keyptr_read_only);
+#endif
+#if R > 5
+    philox_bump(keyptr); philox_round(ctrptr, keyptr_read_only);
+#endif
+#if R > 6
+    philox_bump(keyptr); philox_round(ctrptr, keyptr_read_only);
+#endif
+#if R > 7
+    philox_bump(keyptr); philox_round(ctrptr, keyptr_read_only);
+#endif
+#if R > 8
+    philox_bump(keyptr); philox_round(ctrptr, keyptr_read_only);
+#endif
+#if R > 9
+    philox_bump(keyptr); philox_round(ctrptr, keyptr_read_only);
+#endif
+#if R > 10
+    philox_bump(keyptr); philox_round(ctrptr, keyptr_read_only);
+#endif
+#if R > 11
+    philox_bump(keyptr); philox_round(ctrptr, keyptr_read_only);
+#endif
+#if R > 12
+    philox_bump(keyptr); philox_round(ctrptr, keyptr_read_only);
+#endif
+#if R > 13
+    philox_bump(keyptr); philox_round(ctrptr, keyptr_read_only);
+#endif
+#if R > 14
+    philox_bump(keyptr); philox_round(ctrptr, keyptr_read_only);
+#endif
+#if R > 15
+    philox_bump(keyptr); philox_round(ctrptr, keyptr_read_only);
+#endif
```

Comment:
@kaatish I am guessing, the bump is incrementing something for each of those `#if` independent of one another, is that right ? Can't we put this in a loop ? I see that R is a macro for value `PHILOX_DEFAULT_ROUNDS`. I don't think we are going to get much of performance difference, can you try it with loop and see how much time is it taking for a random example ?